components/service-waiter/cmd/database.go

```diff
@@ -65,52 +82,71 @@ DB_CA_CERT and DB_USER(=gitpod)`,
+			Timestamp: viper.GetString("migration-timestamp"),
+		}
+
+		log.WithField("timeout", timeout.String()).WithField("dsn", censoredDSN).WithField("migrationName", migration.Name).WithField("migrationTimestamp", migration.Timestamp).Info("waiting for database")
+		for ctx.Err() == nil {
+			log.Info("attempting to check if database is available")
+			if err := checkDbAvailable(ctx, cfg, &migration); err != nil {
+				log.WithError(err).Debug("retry")
+				<-time.After(time.Second)
+			} else {
 				break
 			}
+		}
 
-			done <- true
-		}()
-
-		select {
-		case <-done:
+		if ctx.Err() != nil {
+			log.WithField("timeout", timeout.String()).WithError(ctx.Err()).Fatal("database did not become available in time")
+		} else {
 			log.Info("database became available")
-			return
-		case <-time.After(timeout):
-			log.WithField("timeout", timeout.String()).Fatal("database did not become available in time")
 		}
 	},
 }
 
+// checkDbAvailable will connect and check if migrations table contains the latest migration
+func checkDbAvailable(ctx context.Context, cfg *mysql.Config, migration *migrationInfo) error {
+	db, err := sql.Open("mysql", cfg.FormatDSN())
+	if err != nil {
+		return err
+	}
+	// ignore error
+	defer db.Close()
+
+	// if migration info is not set, just ping the database
+	if migration.Name == "" || migration.Timestamp == "" {
+		return db.PingContext(ctx)
+	}
+
+	log.Info("checking if database is migrated")
+	row := db.QueryRowContext(ctx, "SELECT name FROM "+migrationTableName+" WHERE timestamp = ?", migration.Timestamp)
+	var name string
+	if err := row.Scan(&name); err != nil {
+		log.WithError(err).Error("failed to fetch related migration")
+		return err
+	}
+	// migration name should be `{migrationName}{timestamp}`. i.e. AddGitStatusColumnToWorkspaceInstance1690915807191
+	if name != fmt.Sprintf("%s%s", migration.Name, migration.Timestamp) {
+		log.WithError(err).WithField("latest", name).WithField("checking", migration).Error("migrations table does not contain the latest migration")
+		return err
+	}
+
+	return nil
+}
+
 func init() {
 	rootCmd.AddCommand(databaseCmd)
 
+	// Database config
 	databaseCmd.Flags().StringP("host", "H", os.Getenv("DB_HOST"), "Host to try and connect to")
 	databaseCmd.Flags().StringP("port", "p", envOrDefault("DB_PORT", "3306"), "Port to connect on")
 	databaseCmd.Flags().StringP("password", "P", os.Getenv("DB_PASSWORD"), "Password to use when connecting")
 	databaseCmd.Flags().StringP("username", "u", envOrDefault("DB_USERNAME", "gitpod"), "Username to use when connected")
 	databaseCmd.Flags().StringP("caCert", "", os.Getenv("DB_CA_CERT"), "Custom CA cert (chain) to use when connected")
+
+	// Database migration config
+	databaseCmd.Flags().StringP("migration-name", "", os.Getenv("DB_MIGRATION_NAME"), "The latest migration name")
+	databaseCmd.Flags().StringP("migration-timestamp", "", os.Getenv("DB_MIGRATION_TIMESTAMP"), "The latest migration timestamp")
```

Comment:
I don't think we need to check the timestamp or does Typeorm do?

Reply:
name in database is something like `AddGitStatusColumnToWorkspaceInstance1690915807191`

<img width="1038" alt="image" src="https://github.com/gitpod-io/gitpod/assets/20944364/7d0b5847-6364-43a8-8309-e7b3f04b4d51">

Reply:
That is the name of the class

Reply:
Maybe I can change sql query into `SELECT * FROM migrations WHERE name=? LIMIT 1`, and pass `formatted name`. But we still need both name and timestamp or do format name inside `print-migration.ts`

Reply:
I'll take a closer look at the PR in a few minutes

Reply:
We can change `print-migration.ts`  into code below then we can get latest migration name, will look into it later
```
const config = new Config();
    const typeorm = new TypeORM(config, {});
    const conn = await typeorm.getConnection();
    const t = await new MigrationExecutor(conn).getAllMigrations();
```